sql: mark functions as Immutable/Stable/Volatile

otan opened the pull request:
This commit adds the "Volatile" argument to all function properties,
replacing the "Impure" definition. I've replaced all "Impure" checks
with "!= VolatilityImmutable` to give equivalent functionality for now -
we can enhance the folding and the like in a later commit. The planner
tweaks can come from someone more familiar with it :).

Also updated the pg_proc table, which should be the only visible user
change.

Refs #26582.

Release note (sql change): Populated the `pg_proc` table's `provolatile`
field based on the internal builtin volatility definition. This value
used to always be NULL.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/48491)
<!-- Reviewable:end -->

otan commented:
(let me know if this should be "touches" instead)

mjibson commented on pkg/sql/pg_catalog.go:
```diff
@@ -2179,6 +2179,19 @@ CREATE TABLE pg_catalog.pg_proc (
+						case tree.VolatilityVolatile:
+							volatile = 'v'
+						default:
+							return errors.Newf("unexpected volatility %d on proname %s", props.Volatility, name)
```
We should probably panic here. If it's something we don't recognize, that indicates a more major problem.

mjibson commented on pkg/sql/sem/tree/function_definition.go:
```diff
@@ -28,6 +28,30 @@ type FunctionDefinition struct {
+const (
+	// VolatilityImmutable indicates the builtin result never changes for a
+	// given input.
+	VolatilityImmutable Volatility = iota
```
I think we should use the characters here. That eases the stringification and also makes the zero value something that should never happen (and thus always incorrect if it does).

otan replied:
fine with me.

mjibson commented on pkg/sql/sem/tree/function_definition.go:
```diff
@@ -28,6 +28,30 @@ type FunctionDefinition struct {
+// on the values of its explicit arguments, or can change due to outside factors
+// (such as parameter variables or table contents).
+//
+// Each value here must be ordered such that the property i+1 is "weaker" the property
```
Why does the order matter? I don't see any code that cares about the relative ordering of this.

otan replied:
ah, i was planning to actually use it then decided not to.. resolved.

madelynnblue reviewed (commented):
I would like a test added that asserts that all for all builtins with fnProps defined, the volatility is not the zero value.

otan commented:
I've made the builtins all take immutable (like we do today) if not filled in on init time. Do you want that to be different?

otan commented:
(would you rather make fnProps fill in the volatility by default, and then anyone who defined FunctionProperties manually must define their own?)

mjibson commented on pkg/sql/sem/builtins/all_builtins.go:
```diff
@@ -56,9 +56,12 @@ func init() {
 	for _, name := range AllBuiltinNames {
 		def := builtins[name]
+		if def.props.Volatility == 0 {
+			def.props.Volatility = tree.VolatilityImmutable
```
Can you add something like: `TODO: instead of the default being immutable (which could be a problem if a new function forgets to set its volatility if not immutable), explicitly define volatility of all functions and add a test that asserts it is never the zero value.`

RaduBerinde commented:
> I've made the builtins be defined as immutable (like we do today as impure=false) if not filled in on init time. Do you want that to be different?

Unfortunately, what we currently do is the worst thing we can do.. If someone adds a builtin without thinking about volatility or even knowing much about it, we should be conservative with the default. I do prefer that we require it explicitly everywhere, and maybe just document that "if in doubt, use volatile". I realize this is a very tedious change :/

Note that because of this unsafe default, we can't completely trust this flag today and the optimizer has its own "whitelist" of functions here: https://github.com/cockroachdb/cockroach/blob/2f08e8ea772d4619b66182020cb9453ad341fcbb/pkg/sql/opt/norm/fold_constants.go#L418
This would need to be resolved before closing the issue.

otan commented:
alright, we can make that change separately. I'll put this in as a first step.

otan committed:
sql: mark functions as Immutable/Stable/Volatile

This commit adds the "Volatile" argument to all function properties,
replacing the "Impure" definition. I've replaced all "Impure" checks
with "!= VolatilityImmutable` to give equivalent functionality for now -
we can enhance the folding and the like in a later commit. The planner
tweaks can come from someone more familiar with it :).

Also updated the pg_proc table, which should be the only visible user
change.

Release note (sql change): Populated the `pg_proc` table's `provolatile`
field based on the internal builtin volatility definition. This value
used to always be NULL.

mjibson commented:
Is this ready to go now?

otan commented:
yep!

i'm planning on auditing everything for my flex friday, and doing the overload approach (unless you had something in mind). may be some more refactors over the same file, but whatever :P

bors r=mjibson

mjibson commented:
Thanks, that sounds great.

RaduBerinde reviewed (commented):
Argh..... I had some comments but apparently never hit Publish.. sorry about that. Can be addressed in a follow-up PR

Nice, thanks for working on this! CC @rytaft @andy-kimball 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48491#-:-M6eXDbs9U1wu-SEQ780:bi01mf)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mjibson, @otan, and @RaduBerinde)

___
*[pkg/sql/sem/tree/function_definition.go, line 43 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/48491#-M6eXata6e7BvKYHnzIm:-M6eXata6e7BvKYHnzIn:b-m3ocpo) ([raw file](https://github.com/cockroachdb/cockroach/blob/ba3a3ab5ddb33626e3125883709577f023f77d64/pkg/sql/sem/tree/function_definition.go#L43)):*
> ```Go
> type Volatility int
> 
> const (
> ```

We need to document these as clearly as possible because this will be consulted any time someone adds or messes with a builtin, and it will be used when figuring out the correct logic in the optimizer. Better to be extra verbose - we should mention exactly what we can do with some examples. A while back I had thought of what semantics we would want for this property on the optimizer side, I wrote this comment: https://github.com/RaduBerinde/cockroach/blob/c11ad89fe595f9d378e234691bbd1f7d42eab488/pkg/sql/opt/props/volatility.go

We can lift some descriptions / comments from there. We should consider separating "volatile" from "mutation" (or "side-effecting") at the function level.
___
*[pkg/sql/sem/tree/function_definition.go, line 45 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/48491#-M6eXn7S29GjL1KvNza7:-M6eXn7S29GjL1KvNza8:b-ul0qfn) ([raw file](https://github.com/cockroachdb/cockroach/blob/ba3a3ab5ddb33626e3125883709577f023f77d64/pkg/sql/sem/tree/function_definition.go#L45)):*
> ```Go
> const (
> 	// VolatilityImmutable indicates the builtin result never changes for a
> 	// given input.
> ```

It does not depend on configuration settings, and does not modify the database, the transaction state, or any other state. Immutable functions can always be constant-folded (even in cached plans).
___
*[pkg/sql/sem/tree/function_definition.go, line 48 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/48491#-M6eVM4-C_Gj_YxF4OVn:-M6eVM4-C_Gj_YxF4OVo:b-vwpfiu) ([raw file](https://github.com/cockroachdb/cockroach/blob/ba3a3ab5ddb33626e3125883709577f023f77d64/pkg/sql/sem/tree/function_definition.go#L48)):*
> ```Go
> 	VolatilityImmutable Volatility = iota
> 	// VolatilityStable indicates the builtin result never changes during
> 	// a given scan.
> ```

"during a given scan" is not very helpful. Maybe change to "the function cannot modify the database or the transaction state and is guaranteed to return the same results given the same arguments whenever it is evaluated within the same statement."
___
*[pkg/sql/sem/tree/function_definition.go, line 50 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/48491#-M6eVtPI0l2bNrQhXkkb:-M6eVtPI0l2bNrQhXkkc:bfd9542) ([raw file](https://github.com/cockroachdb/cockroach/blob/ba3a3ab5ddb33626e3125883709577f023f77d64/pkg/sql/sem/tree/function_definition.go#L50)):*
> ```Go
> 	// a given scan.
> 	VolatilityStable
> 	// VolatilityVolatile indicates the builtin result can change even
> ```

mention "or that it has side effects". I would remove the "impure" thing because it doesn't help here without a proper definition; we can add some examples of different things the function can do.
___
*[pkg/sql/sqlbase/table.go, line 34 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48491#-M6ebZcC4VMWlRh1dHfY:-M6ebZcC4VMWlRh1dHfZ:b-x69j9h) ([raw file](https://github.com/cockroachdb/cockroach/blob/198edc0f58f4e2c1b20a6cd6164f17c8cd135583/pkg/sql/sqlbase/table.go#L34)):*
> ```Go
> // type and contains no variable expressions. It returns the type-checked and
> // constant-folded expression.
> func SanitizeVarFreeExpr(
> ```

[nit] I would break this out into `SanitizeVarFreeExpr` and `SanitizeImmutableExpr`. They can call out to the same unexported function (maybe passing the `flags` directly)


<!-- Sent from Reviewable.io -->

andy-kimball commented on pkg/sql/sem/tree/function_definition.go:
```diff
@@ -28,6 +28,29 @@ type FunctionDefinition struct {
+	// given input.
+	VolatilityImmutable Volatility = 'i'
+	// VolatilityStable indicates the builtin result never changes during
+	// a given scan.
```
The "given scan" part confuses me. I was expecting something like:
```
Immutable: never changes for a given input, even across multiple executions of a statement
Stable: never changes for a given input when invoked multiple times in a single statement, but may change across multiple executions of a statement
Volatile: may change when invoked multiple times in a single statement
```
Examples of each case would be helpful as well, to give people a quick intuition:
```
Immutable: length(col)
Stable: now()
Volatile: random()
```

otan replied:
I'll fix this up tomorrow! I have 1000 functions to audit :D
